Add option for Transform on Request to AppNote0022 (#182)
bbc/tams: docs/appnotes/0022-transform-architectures.md

* Adds control signalling to the TAMS API
* Potentially impedes more complex configuration of transcoders
* More components may mean a more complicated architecture

## Transcode on Request

This architecture sees transcode of media being triggered just-in-time when a consuming client attempts to read media segments in a format not currently available on-disk.

> [!NOTE]
> This option likely requires deep integration into the TAMS Service due to the potential creation and removal of Media Object Instances.

The requesting client will create a new Flow against the existing Source with the required codec and properties, and the `trigger_transcode` tag set to `virtual`.

> The TAMS Service will populate Segments and Objects correlating with the timeline range covered by the original Segments.

Reviewer: I'm trying to hint here that you might implement this without creating anything durable: instead rewriting the response of a database query to figure out where the segment boundaries are. But I'm not sure my attempt is any better!
```suggestion
From a client's perspective, Segments and Objects are visible in the new Flow corresponding with the   timerange covered by the original Flow.
```

Author: I think my concern with this wording is while it captures the apparent behaviour, it doesn't explicitly describe anything actionable. We could add a further sentence along the lines of "How the Service populates these Segments is up to the implementation". But that also feels a touch clunky.

Reviewer: Ok, let's leave this as-is then

These Objects will have instances that are "virtual".
That is to say that the Objects do not exist on disk.
When a GET request is performed with the instance URL, the Service will perform the transcode to create the segment and return it, as if the media existed on disk.

Where the originating Flow has a `flow_status` of `ingesting`, new virtual Segments should be created as new originating Segments are registered.

### Pros

> * Requesting clients integrate with fewer APIs

Reviewer: Small addition to call out the Con this mitigates from above
```suggestion
* Transcode only takes place when a client attempts to use the transcoded media
* Requesting clients integrate with fewer APIs
```

* Transcode capability to be deployed in close (logical) proximity to the rest of the TAMS Service, due to its deep integration with it
* Fewer components potentially means a less complicated architecture overall

### Cons

* Adds control signalling to the TAMS API
* Potentially impedes more complex configuration of transcoders

> * Deep integration adds more complexity to TAMS Service implementations

Reviewer: ```suggestion
* Deep integration adds more complexity to TAMS Service implementations
* Segments need to be short enough and transcode fast enough to respond to Object GET in a timely manner
```

## Transcode on Request

> This architecture sees transcode/transpackage of media being triggered just-in-time when a consuming client attempts to read media segments in a format not currently available on-disk.

Reviewer: Nit: Given our discussion regarding #181, is "transformation" a better word to use here?

Author: As per the discussion in the TSC Slack, I've left this as transcode/transpackage as the signalling method only really supports those types of operation. I added the extension to call out that the Virtual Objects approach may be applied to other forms of transform, though. But would require a different approach to configuration.
